File: swift/Sources/MDoc/MdocPresenter.swift
import CborCose
import Foundation
import WalletAPI

/// How the mdoc authenticates its response (ISO 18013-5 §9.1.3.5). A wallet whose `DeviceKey` is a
/// signing key produces a `deviceSignature`; one whose `DeviceKey` is a key-agreement key produces a
/// `deviceMac` instead. Readers must accept either — see `MdocReader.verifyDeviceResponse`.
public enum DeviceAuth {
    /// `deviceSignature`: an ECDSA COSE_Sign1 over `DeviceAuthentication`.
    case signature(signer: any CoseSigner, algorithm: SigningAlgorithm = .es256)

    /// `deviceMac`: HMAC-256/256 over `DeviceAuthentication` keyed by the `EMacKey` — HKDF over the ECDH
    /// secret of the mdoc's `DeviceKey` and the reader's `EReaderKey`, salted by the SessionTranscript.
    /// The mdoc computes that secret inside its secure area, so the MAC proves possession of `DeviceKey`
    /// to *this reader only* — unlike a signature, it is not transferable to a third party.
    case mac(emacKey: [UInt8])
}

/// One document of a (possibly multi-document) `DeviceResponse`: which issuer-signed credential to
/// present, which elements to disclose, and how the device authenticates it. Each document carries
/// its own `DeviceAuth` — `DeviceAuthentication` is per-docType and each mdoc has its own DeviceKey.
public struct PresentedDocument {
    public let issuerSigned: IssuerSigned
    public let docType: String
    /// namespace -> element identifiers to disclose.
    public let disclosed: [String: [String]]
    public let deviceAuth: DeviceAuth
    /// Device-signed data elements (namespace -> id -> value), e.g. OpenID4VP mdoc transaction data (B.2.1).
    public let deviceSignedNamespaces: [String: [String: Cbor]]

    public init(
        issuerSigned: IssuerSigned,
        docType: String,
        disclosed: [String: [String]],
        deviceAuth: DeviceAuth,
        deviceSignedNamespaces: [String: [String: Cbor]] = [:]
    ) {
        self.issuerSigned = issuerSigned
        self.docType = docType
        self.disclosed = disclosed
        self.deviceAuth = deviceAuth
        self.deviceSignedNamespaces = deviceSignedNamespaces
    }
}

/// Builds an mdoc `DeviceResponse` (ISO 18013-5 §8.3.2.1.2.2) for presentation: keeps only the
/// disclosed issuer-signed items and produces `DeviceSigned` — a `deviceSignature` or a `deviceMac`
/// over the `DeviceAuthentication` structure (detached payload) bound to the `sessionTranscript`.
public enum MdocPresenter {

    private static let tagEncodedCbor: UInt64 = 24

    /// Convenience for the signature form; equivalent to passing `.signature`.
    public static func deviceResponse(
        issuerSigned: IssuerSigned,
        docType: String,
        disclosed: [String: [String]],
        sessionTranscript: Cbor,
        deviceSigner: any CoseSigner,
        deviceSignAlgorithm: SigningAlgorithm = .es256
    ) async throws -> [UInt8] {
        try await deviceResponse(
            issuerSigned: issuerSigned, docType: docType, disclosed: disclosed,
            sessionTranscript: sessionTranscript,
            deviceAuth: .signature(signer: deviceSigner, algorithm: deviceSignAlgorithm))
    }

    /// Convenience for the single-document response (one `DocRequest` answered).
    public static func deviceResponse(
        issuerSigned: IssuerSigned,
        docType: String,
        disclosed: [String: [String]], // namespace -> element identifiers to disclose
        sessionTranscript: Cbor,
        deviceAuth: DeviceAuth,
        /// Device-signed data elements (namespace -> id -> value), e.g. OpenID4VP mdoc transaction data (B.2.1).
        deviceSignedNamespaces: [String: [String: Cbor]] = [:]
    ) async throws -> [UInt8] {
        try await deviceResponse(
            documents: [PresentedDocument(
                issuerSigned: issuerSigned, docType: docType, disclosed: disclosed,
                deviceAuth: deviceAuth, deviceSignedNamespaces: deviceSignedNamespaces)],
            sessionTranscript: sessionTranscript)
    }

    /// A `DeviceResponse` answering several `DocRequest`s at once — one `Document` per `documents` entry.
    public static func deviceResponse(documents: [PresentedDocument], sessionTranscript: Cbor) async throws -> [UInt8] {
        guard !documents.isEmpty else { throw MdocError("a DeviceResponse needs at least one document") }
        var docs: [Cbor] = []
        for presented in documents {
            docs.append(try await document(presented, sessionTranscript: sessionTranscript))
        }
        let deviceResponse = Cbor.map([
            (.text("version"), .text("1.0")),
            (.text("documents"), .array(docs)),
            (.text("status"), .int(0)),
        ])
        return try CborEncoder.encode(deviceResponse)
    }

    private static func document(_ presented: PresentedDocument, sessionTranscript: Cbor) async throws -> Cbor {
        let issuerSigned = presented.issuerSigned
        let docType = presented.docType
        let disclosed = presented.disclosed
        let deviceAuth = presented.deviceAuth
        let deviceSignedNamespaces = presented.deviceSignedNamespaces
        // Keep only the disclosed items, re-emitting their exact IssuerSignedItemBytes (#6.24).
        var filteredNs: [(Cbor, Cbor)] = []
        for (ns, items) in issuerSigned.nameSpaces {
            guard let ids = disclosed[ns] else { continue }
            let kept = try items.filter { ids.contains($0.item.elementIdentifier) }.map { try CborDecoder.decode($0.itemBytes) }
            if !kept.isEmpty { filteredNs.append((.text(ns), .array(kept))) }
        }
        let issuerSignedCbor = Cbor.map([
            (.text("nameSpaces"), .map(filteredNs)),
            (.text("issuerAuth"), issuerSigned.issuerAuth.toCbor(tagged: false)),
        ])

        // DeviceNameSpaces: empty for a basic presentation, or the provided device-signed data elements.
        let deviceNsMap = Cbor.map(deviceSignedNamespaces.map { (ns, elements) in
            (Cbor.text(ns), Cbor.map(elements.map { (Cbor.text($0.key), $0.value) }))
        })
        let deviceNameSpacesBytes = Cbor.tagged(tagEncodedCbor, .bytes(try CborEncoder.encode(deviceNsMap)))

        let deviceAuthentication = Cbor.array([.text("DeviceAuthentication"), sessionTranscript, .text(docType), deviceNameSpacesBytes])
        let deviceAuthBytes = try CborEncoder.encode(.tagged(tagEncodedCbor, .bytes(try CborEncoder.encode(deviceAuthentication))))

        let deviceAuthEntry: (Cbor, Cbor)
        switch deviceAuth {
        case let .signature(signer, algorithm):
            let deviceSignature = try await CoseSign1.sign(
                protected: CoseHeaders.of(algorithm: algorithm.coseAlgorithm),
                payload: nil,
                detachedPayload: deviceAuthBytes,
                signer: signer)
            deviceAuthEntry = (.text("deviceSignature"), deviceSignature.toCbor(tagged: false))
        case let .mac(emacKey):
            let deviceMac = try CoseMac0.create(key: emacKey, detachedPayload: deviceAuthBytes)
            deviceAuthEntry = (.text("deviceMac"), deviceMac.toCbor(tagged: false))
        }

        let deviceSigned = Cbor.map([
            (.text("nameSpaces"), deviceNameSpacesBytes),
            (.text("deviceAuth"), .map([deviceAuthEntry])),
        ])

        return Cbor.map([
            (.text("docType"), .text(docType)),
            (.text("issuerSigned"), issuerSignedCbor),
            (.text("deviceSigned"), deviceSigned),
        ])
    }
}

File: kotlin/mdoc/src/main/kotlin/com/hopae/eudi/wallet/mdoc/MdocPresenter.kt
package com.hopae.eudi.wallet.mdoc

import com.hopae.eudi.wallet.cbor.Cbor
import com.hopae.eudi.wallet.cbor.CborDecoder
import com.hopae.eudi.wallet.cbor.CborEncoder
import com.hopae.eudi.wallet.cbor.cose.CoseHeaders
import com.hopae.eudi.wallet.cbor.cose.CoseMac0
import com.hopae.eudi.wallet.cbor.cose.CoseSign1
import com.hopae.eudi.wallet.cbor.cose.CoseSigner
import com.hopae.eudi.wallet.spi.SigningAlgorithm
import com.hopae.eudi.wallet.spi.coseAlgorithm

/**
 * How the mdoc authenticates its response (ISO 18013-5 §9.1.3.5). A wallet whose `DeviceKey` is a
 * signing key produces a `deviceSignature`; one whose `DeviceKey` is a key-agreement key produces a
 * `deviceMac` instead. Readers must accept either — see [MdocReader.verifyDeviceResponse].
 */
sealed interface DeviceAuth {
    /** `deviceSignature`: an ECDSA COSE_Sign1 over `DeviceAuthentication`. */
    class Signature(
        val signer: CoseSigner,
        val algorithm: SigningAlgorithm = SigningAlgorithm.ES256,
    ) : DeviceAuth

    /**
     * `deviceMac`: HMAC-256/256 over `DeviceAuthentication` keyed by the `EMacKey` — HKDF over the ECDH
     * secret of the mdoc's `DeviceKey` and the reader's `EReaderKey`, salted by the SessionTranscript.
     * The mdoc computes that secret inside its secure area, so the MAC proves possession of `DeviceKey`
     * to *this reader only* — unlike a signature, it is not transferable to a third party.
     */
    class Mac(val emacKey: ByteArray) : DeviceAuth
}

/**
 * One document of a (possibly multi-document) `DeviceResponse`: which issuer-signed credential to
 * present, which elements to disclose, and how the device authenticates it. Each document carries
 * its own [DeviceAuth] — `DeviceAuthentication` is per-docType and each mdoc has its own DeviceKey.
 */
class PresentedDocument(
    val issuerSigned: IssuerSigned,
    val docType: String,
    /** namespace -> element identifiers to disclose. */
    val disclosed: Map<String, List<String>>,
    val deviceAuth: DeviceAuth,
    /** Device-signed data elements (namespace -> id -> value), e.g. OpenID4VP mdoc transaction data (B.2.1). */
    val deviceSignedNamespaces: Map<String, Map<String, Cbor>> = emptyMap(),
)

/**
 * Builds an mdoc `DeviceResponse` (ISO 18013-5 §8.3.2.1.2.2) for presentation: keeps only the
 * disclosed issuer-signed items and produces `DeviceSigned` — a `deviceSignature` or a `deviceMac`
 * over the `DeviceAuthentication` structure (detached payload) bound to the [sessionTranscript].
 */
object MdocPresenter {

    /** Convenience for the signature form; equivalent to passing [DeviceAuth.Signature]. */
    suspend fun deviceResponse(
        issuerSigned: IssuerSigned,
        docType: String,
        disclosed: Map<String, List<String>>,
        sessionTranscript: Cbor,
        deviceSigner: CoseSigner,
        deviceSignAlgorithm: SigningAlgorithm = SigningAlgorithm.ES256,
    ): ByteArray = deviceResponse(
        issuerSigned, docType, disclosed, sessionTranscript,
        DeviceAuth.Signature(deviceSigner, deviceSignAlgorithm),
    )

    /** Convenience for the single-document response (one `DocRequest` answered). */
    suspend fun deviceResponse(
        issuerSigned: IssuerSigned,
        docType: String,
        /** namespace -> element identifiers to disclose. */
        disclosed: Map<String, List<String>>,
        sessionTranscript: Cbor,
        deviceAuth: DeviceAuth,
        /** Device-signed data elements (namespace -> id -> value), e.g. OpenID4VP mdoc transaction data (B.2.1). */
        deviceSignedNamespaces: Map<String, Map<String, Cbor>> = emptyMap(),
    ): ByteArray = deviceResponse(
        listOf(PresentedDocument(issuerSigned, docType, disclosed, deviceAuth, deviceSignedNamespaces)),
        sessionTranscript,
    )

    /** A `DeviceResponse` answering several `DocRequest`s at once — one `Document` per [documents] entry. */
    suspend fun deviceResponse(documents: List<PresentedDocument>, sessionTranscript: Cbor): ByteArray {
        require(documents.isNotEmpty()) { "a DeviceResponse needs at least one document" }
        val deviceResponse = Cbor.CborMap(
            listOf(
                Cbor.Text("version") to Cbor.Text("1.0"),
                Cbor.Text("documents") to Cbor.Array(documents.map { document(it, sessionTranscript) }),
                Cbor.Text("status") to Cbor.int(0),
            )
        )
        return CborEncoder.encode(deviceResponse)
    }

    private suspend fun document(presented: PresentedDocument, sessionTranscript: Cbor): Cbor {
        val issuerSigned = presented.issuerSigned
        val docType = presented.docType
        val disclosed = presented.disclosed
        val deviceAuth = presented.deviceAuth
        val deviceSignedNamespaces = presented.deviceSignedNamespaces
        // Keep only the disclosed items, re-emitting their exact IssuerSignedItemBytes (#6.24).
        val filteredNs = issuerSigned.nameSpaces.mapNotNull { (ns, items) ->
            val ids = disclosed[ns] ?: return@mapNotNull null
            val kept = items.filter { it.item.elementIdentifier in ids }
            if (kept.isEmpty()) null else Cbor.Text(ns) to Cbor.Array(kept.map { CborDecoder.decode(it.itemBytes) })
        }
        val issuerSignedCbor = Cbor.CborMap(
            listOf(
                Cbor.Text("nameSpaces") to Cbor.CborMap(filteredNs),
                Cbor.Text("issuerAuth") to issuerSigned.issuerAuth.toCbor(tagged = false),
            )
        )

        // DeviceNameSpaces: empty for a basic presentation, or the provided device-signed data elements.
        val deviceNsMap = Cbor.CborMap(
            deviceSignedNamespaces.map { (ns, elements) ->
                Cbor.Text(ns) to Cbor.CborMap(elements.map { (id, value) -> Cbor.Text(id) to value })
            }
        )
        val deviceNameSpacesBytes = Cbor.Tagged(TAG_ENCODED_CBOR, Cbor.Bytes(CborEncoder.encode(deviceNsMap)))

        // DeviceAuthentication = ["DeviceAuthentication", SessionTranscript, DocType, DeviceNameSpacesBytes]
        val deviceAuthentication =
            Cbor.Array(listOf(Cbor.Text("DeviceAuthentication"), sessionTranscript, Cbor.Text(docType), deviceNameSpacesBytes))
        val deviceAuthBytes = CborEncoder.encode(Cbor.Tagged(TAG_ENCODED_CBOR, Cbor.Bytes(CborEncoder.encode(deviceAuthentication))))

        val deviceAuthEntry = when (deviceAuth) {
            is DeviceAuth.Signature -> Cbor.Text("deviceSignature") to CoseSign1.sign(
                protected = CoseHeaders.of(algorithm = deviceAuth.algorithm.coseAlgorithm),
                payload = null,
                detachedPayload = deviceAuthBytes,
                signer = deviceAuth.signer,
            ).toCbor(tagged = false)

            is DeviceAuth.Mac -> Cbor.Text("deviceMac") to CoseMac0.create(
                key = deviceAuth.emacKey,
                detachedPayload = deviceAuthBytes,
            ).toCbor(tagged = false)
        }

        val deviceSigned = Cbor.CborMap(
            listOf(
                Cbor.Text("nameSpaces") to deviceNameSpacesBytes,
                Cbor.Text("deviceAuth") to Cbor.CborMap(listOf(deviceAuthEntry)),
            )
        )

        return Cbor.CborMap(
            listOf(
                Cbor.Text("docType") to Cbor.Text(docType),
                Cbor.Text("issuerSigned") to issuerSignedCbor,
                Cbor.Text("deviceSigned") to deviceSigned,
            )
        )
    }
}
